### DISCUSSION TITLE
Already existing Rules used in internal projects that could become part of BC.LinterCop

### DISCUSSION THREAD

**Discussion by tinestaric** [StefanMaron/BusinessCentral.LinterCop#269] [New Rule]
Hey,

at one of the earlier points of this repo, I forked out and started developing custom code cop rules that we needed for the project. It has since developed into quite a collection of rules, and now I think some of them might be of interest to a broader community that's already using BC.LinterCop. 

I'll list all that we have, and if any seem of interest, I can prepare a PR to include them in the BC.LinterCop.

1. Global Var, Local Var, Parameter, and Return Value Prefix 
used in projects that still want Hungarian notation. Enforces variable prefixing, which could be used with LinterCop.json to allow custom prefixes instead of "g", "l", and "p". Turned off by default, as the majority of projects don't use Hungarian notation anymore

2. If statements need to be encased with brackets
Fairly project-specific request, but maybe there's anyone else who does this?

3. Semicolon after procedure definition
This is more for consistency's sake than an actual development improvement. The procedure's signature can end with a semicolon or not, it makes no difference, so we added a rule to never have semicolons at the end.

4. Procedures must be either Local or Internal
This one's quite useful for AppSource projects that we want to slowly open up when there's a need for it. It helps as the majority of Public procedures can then be changed without obsoleting them.

5. Object Access and Extensibility Property must be set to Internal
For the same reasons as above, it helps with the obsoletion process.

6. Tool tip punctuation
For consistency's sake, we added a rule that ToolTips must end with a dot.

7. Global Var Section, Trigger Section, and Procedure Section Position
For better readability of objects, we started enforcing that Global Var Section must be above triggers, and triggers must be above Procedures. I know the default MS order is Trigger->GVar->Procedure so the rule could be adjusted to reflect that order instead if that's the preferred way.

8. Label Punctuation
For consistency, we have a rule that Labels that end with Msg or Err must end with a dot, and Qst labels must end with a question mark

9. Internal Methods must have explicit parameters
We had issues where some of the bugs appeared because the developer didn't specify a parameter with .Insert() or .Modify(), which in turn made it skip the trigger. This rule enforces explicit definitions of parameters for Insert, Modify, Delete, Update, and DeleteAll internal methods.

10. Empty Object Sections
With the default code snippets, some objects get empty sections that developers forgot to remove, so we added a rule to check for empty OnRun triggers on Codeunits, and empty Area and Actions sections on Pages and PageExtensions

11. Empty Captions should be Locked
Translation files had a lot of unnecessary sections for Enum Captions that were empty (' '), so we added a rule that empty captions should be locked. Right now it only checks through Enum values as that was the only place where we were using empty captions.

12. SetAutoCalcFields on Normal Fields
A rule that checks if the fields used in the SetAutoCalcFields function are flow fields, as it will produce a run-time error when trying to execute the line. This was also reported to MS but was closed with "Not planned" as they have a big overhaul of Code Analysis in plans. It will probably be obsolete when they do the update, but we still needed a check until that happens.

13. Zero Enum Values Reserved
Quite often we got a request to add an empty value to the existing enum and place it at the start. We added a rule that the 0 enum value should be reserved for zero values, so this kind of request can be easily handled.

14. Tok Labels need to be locked
More often than not label with a Tok suffix was used for technical constants, so we added a rule to require that these labels are Locked.

15. Procedure Name, Variables, and Parameters must not contain whitespace or Wildcard symbols
self-explanatory

16. First 19 fields are reserved for Primary Keys fields
It helps with the readability of the tables, as primary key fields are instantly visible

17. Captions are required on Fields, Parts, Groups, Actions, etc...

**Comment by rvanbekkum** [comment]
1. We have implemented a similar rule, so it's not that weird. ;)
2. We have disabled Microsoft's rule that says they should not. I think enforcing a `begin`-`end` pair always might be too specific. For AL it is also less of an issue due to the auto-format on save, but as long as the AL formatter works properly.
3. This is more of a preference indeed. I tend NOT to use a `;` after the procedure declaration.
4. The thing I do agree on is that setting things to internal vs. public vs. private/local should be a though-through decision. But personally I don't think everything should be internal, as sometimes you actually want people to be able to extend and use your procedures.
5. Same as with 4.
6. Simple, but good one.
7. Good one.
8. Simple, but good one.
9. I think same as this one: https://github.com/StefanMaron/BusinessCentral.LinterCop/discussions/111 ?
10. Simple, but good one.
11. Good one.
12. Good one.
13. Very good one. Sometimes you actually might not want to use the `0` value at all, so that you know if the option was set explicitly or not.
14. Good one.
15. Totally agree. We actually have rules that make sure all names are restricted to alphanumeric and underscore characters.
16. Simple, but good one. I do not see anyone breaking it very quickly though. :)
17. Duplicate of: https://github.com/StefanMaron/BusinessCentral.LinterCop/discussions/121 ?

**Reply by tinestaric** [reply]
I agree that parts of the app need to be extendable, but the reasoning behind the rule was that the default should be that things are internal. When we identify a need to open things up, we disable the rule with a pragma for that specific object/procedure, so it becomes a deliberate decision to expose something. The driver for this was that the obsoletion process is a pain, especially if you're trying to change a function, that is public just because you needed to call it from a different object.

**Comment by christianbraeunlich** [comment]
>     16. First 19 fields are reserved for Primary Keys fields
>         It helps with the readability of the tables, as primary key fields are instantly visible

Nice list, can imagine that some work went into building all of that! I do have a question about the number of fields here. Isn't the max cap. at 16 fields for a PK (https://github.com/microsoft/alguidelines/discussions/213) ?

**Reply by tinestaric** [reply]
you are correct. the reason why we have the rule checking up to 19, is that we wanted to have the first non-PK field start with the ID of 20.

**Reply by christianbraeunlich** [reply]
well oopsie, didn't think about that.. that's much smoother, I like it! *writes down notes*

**Comment by Arthurvdv** [comment]
First of all, impressive list of additional rules and I would LOVE to see the most of them coming over to the LinterCop.

I fully embrace the most of the rules on the exception of rule 1. With the beginning of AL, we stopt using the Hungarian notations.

**Reply by tinestaric** [reply]
Agreed on the Hungarian notation, we don't use it except for a handful of projects where the customer is taking over the maintenance, and they explicitly requested that we use the notation. That's why this rule is turned off by default.
I'm actually curious about how many partners still use the Hungarian notation.

**Comment by tinestaric** [comment]
@StefanMaron as the project owner, I'll wait for your opinion, and then get to creating the PRs for the ones you think would be beneficial to the project.

**Reply by StefanMaron** [reply]
1. Nope
2. Nope
3. Like it
4. Yes but default state disabled
5. Already exists for access property, no need for extensibiliy if internal
6. Cool
7. Cool
8. Cool
9. Yes, please
10. Cool
11. Cool
12. Cool
13. Never had this request, maybe disabled by default
14. Like "Token"? Is this a valid suffix by MS?
15. I would further extend this rule to only alow A-Za-z. Personally I'd also like to see this rule for Object names and Table fields, controls
16. No need I think. The order in the tooltip is not dependend in the ID but the actual order the fields are placed in code
17. Aleady exists.

Note: Apart from some few exceptions I want to publish new rules as infos to not break any pipelines that have fail on warning and the linter cop enabled. So if

**Reply by tinestaric** [reply]
as for 14, yeah, Tok suffix for Tokens. this is an excerpt from AA0074 Label Naming Warning documentation:
<html>
<body>


Three-letter suffix | Meaning
-- | --
Msg | Message
Tok | Token
Err | Error
Qst | StrMenu or Confirm
Lbl | Label, Caption
Txt | Text


</body>
</html>

Sidenote, should the PRs be created per rule, in batches, or as a single PR? What's your preference?

**Reply by StefanMaron** [reply]
Maybe create it in a few batches. I will try to review and release them as prerelease one after another
